src/libsystemd/sd-event/event-util.c

```diff
@@ -274,16 +274,17 @@ int event_forward_signals(
                 return -ENOMEM;
 
         FOREACH_ARRAY(sig, signals, n_signals) {
-                r = sd_event_add_signal(e, &sources[n_sources], *sig | SD_EVENT_SIGNAL_PROCMASK, event_forward_signal_callback, child);
+                sd_event_source *new_source;
+                r = sd_event_add_signal(e, &new_source, *sig | SD_EVENT_SIGNAL_PROCMASK, event_forward_signal_callback, child);
                 if (r < 0)
                         return r;
+                sources[n_sources++] = new_source;
 
-                r = sd_event_source_set_destroy_callback(sources[n_sources], event_forward_signal_destroy);
+                r = sd_event_source_set_destroy_callback(new_source, event_forward_signal_destroy);
                 if (r < 0)
                         return r;
 
                 sd_event_source_ref(child);
-                n_sources++;
```

Comment:
Why not simply increment n_sources before setting the callback?

Reply:
The callback would be added to the next source. Or maybe I misunderstood you.

Reply:
Indeed. I thought incrementing earlier can make the code change smaller, but cannot. Please forget the comment in the above.

But, we usually do something like the following:
```c
_cleanup_(sd_event_source_unrefp) sd_event_source *s = NULL;
r = sd_event_add_signal(e, &s, ...);
if (r < 0)
        return r;

r = sd_event_source_set_destroy_callback(s, ...);
if (r < 0)
        return r;

sources[n_sources++] = TAKE_PTR(s);
```
That is, assign a perfect event source to the array with `TAKE_PTR()`. This is effectively equivalent to your code, but we use such pattern at many many places.